src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,124 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            elements = new Object[initialCapacity];
+        } else {
+            throw new IllegalArgumentException("Illegal initial capacity " + initialCapacity);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        checkSize();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        checkSize();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int length = list.size();
+        if (size + length >= elements.length) {
+            grow(size + length);
+        }
+        for (int i = 0; i < length; i++) {
+            T element = list.get(i);
+            add(element);
+        }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T element = get(index);
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        size--;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int i = 0;
+        for (; i < size; i++) {
+            if (element == elements[i]
+                    || (element != null && element.equals(elements[i]))) {
```

Comment:
```suggestion
            if (element == elements[i]
                    || element != null && element.equals(elements[i])) {
```
this brackets are redundant (read about [Operator Precedence](https://www.programiz.com/java-programming/operator-precedence))